bitcoin/_posts/0000-00-00-csfs-ctv-rekey-symmetry.md

```diff
@@ -0,0 +1,171 @@
+secret key `1` and public key `G`. Because CHECKSIG operations commit to the
+Taproot Annex, these scripts require no special handling for the channel
+parties to require each other to place the settlement transaction hash in the
+annex and therefore make it possible for either party to later reconstruct any
+prior state's script for spending. Without the annex, an APO-based
+implementation would either fall back to the additional signing round-trip, or
+using an `OP_RETURN` to force this data to be visible.
+
+
+## Naive CTV-CSFS solution
+
+Can we just commit to an additional hash using an additional signature?
+```
+channel:
+tr(musig(keyA, keyB), raw(CTV IKEY CSFS VERIFY IKEY CSFS VERIFY <S+1> CLTV))
+update(n):
+tr(musig(keyA, keyB), raw(DEPTH NOTIF <settlement-n-hash> CTV ELSE CTV IKEY CSFS VERIFY IKEY CSFS VERIFY <S+n+1> CLTV ENDIF))
+update-stack:
+<settlement-n-sig> <settlement-n-hash> <update-n-sig> <update-n-hash>
+```
+
+This method is broken because the two signatures are not linked in any way. A
+malicious channel partner can place a mismatched settlement hash, and update
+transaction on chain, preventing their partner who has a valid later update
+from reconstructing the scripts and updating the channel state.
+
+One obvious solution would be to combine the update hash and the settlement
+hash, but since bitcoin lacks a concatenation operator, we cannot do that.
+Recently [@4moonsettler](x.com/4moonsettler) proposed
+[`OP_PAIRCOMMIT`](https://github.com/bitcoin/bips/pull/1699) as an alternative
+for this purpose.
+
+
+## CTV-CSFS delegation solution
+
+Now we come to the new solution that we've developed, which ties the update
+and settlement hashes together by the keys which have signed them. CSFS is
+known to be useful for delegation, so we initially delegate to a rekey:
+```
+script(n):
+DUP TOALT DUP TOALT IKEY CSFS VERIFY CTV 2DUP EQUAL NOT VERIFY ROT SWAP FROMALT CSFS VERIFY FROMALT CSFS VERIFY <S+n+1> CLTV
+channel:
+tr(musig(keyA, keyB), raw(<script(0)>))
+update(n):
+tr(musig(keyA, keyB), raw(DEPTH NOTIF <settlement-n-hash> CTV ELSE <script(n)> ENDIF))
+stack:
+<settlement-n-sig> <update-n-sig> <settlement-n-ctv> <update-n-ctv> <rekey-sig> <rekey>
+```
+
+Here the `rekey` is an ephemeral key either randomly generated or derived
+for each state using something like BIP32 and based on the channel key. What
+matters is that the `rekey` is never used to sign anything other than the two
+messages corresponding to the update and settlement hashes for its state. In
+this way they are only valid together and the correct settlement hash must be
+available for a channel partner to reconstruct the scripts and update the
+channel settlement. One quirk of this solution is that if the two signed items
+are allowed to be equal, a malicious partner can simply place the same update
+hash on the stack with its signature twice, so they must be checked for
+inequality by the script.
+
```

Comment:
Key Laddering extends the rekeying approach shown above to allow recursively
rekeying to an arbitrary number of variables. This allows CSFS to be used without
OP\_CAT to sign over collections of variables to be plugged into a script.


For example, for 5 variables (not optimized, written for clarity):

```
DATASIGS: <sd1> <d1> <sd2> <d2> <sd3> <d3> <sd4> <d4> <sd5> <d5>
stack: DATASIGS + <k5> <s5> <k4> <s4> <k3> <s3> <k2> <s2> <k1> <s1>

program:

\\ First, check that k1 is signed by IKEY
    OVER IKEY CSFSV
    DUP TOALT

// Next, Check that k_i signs k_{i+1}
    // stack: DATASIGS + <k5> <s5> <k4> <s4> <k3> <s3> <k2> <s2> <k1>
    // altstack: <k1>

        3DUP ROT SWAP CSFSV 2DROP DUP TOALT

    // stack: DATASIGS + <k5> <s5> <k4> <s4> <k3> <s3> <k2>
    // altstack: <k1> <k2>

        3DUP ROT SWAP CSFSV 2DROP DUP TOALT

    // stack: DATASIGS + <k5> <s5> <k4> <s4> <k3>
    // altstack: <k1> <k2> <k3>

        3DUP ROT SWAP CSFSV 2DROP DUP TOALT

    // stack: DATASIGS + <k5> <s5> <k4>
    // altstack: <k1> <k2> <k3> <k4>

        3DUP ROT SWAP CSFSV 2DROP

    // stack: <sd1> <d1> <sd2> <d2> <sd3> <d3> <sd4> <d4> <sd5> <d5> <k5>
    // altstack: <k1> <k2> <k3> <k4>


        FROMALT FROMALT FROMALT FROMALT

    // stack: <sd1> <d1> <sd2> <d2> <sd3> <d3> <sd4> <d4> <sd5> <d5> <k5> <k4> <k3> <k2> <k1>
    // altstack:

// Now, check each signature of the data

    <6> PICK // sd5
    <6> PICK // d5
    <6> PICK // k5
    CSFSV

    <8> PICK // sd4
    <8> PICK // d4
    <5> PICK // k4
    CSFSV

    <10> PICK // sd3
    <10> PICK // d3
    <4> PICK // k3
    CSFSV

    <12> PICK // sd2
    <12> PICK // d2
    <3> PICK // k2
    CSFSV

    <14> PICK // sd1
    <14> PICK // d1
    <2> PICK // k1
    CSFSV

// Now, Check the inequalities that no key is used as data:
    // stack: <sd1> <d1> <sd2> <d2> <sd3> <d3> <sd4> <d4> <sd5> <d5> <k5> <k4> <k3> <k2> <k1>
    // altstack:

    // No need to check k1 != d0 since no d0

    // Check that k2 != d1
    <1> PICK
    <14> PICK
    NOT EQUAL VERIFY

    // Check that k3 != d2
    <2> PICK
    <12> PICK
    NOT EQUAL VERIFY


    // Check that k4 != d3
    <3> PICK
    <8> PICK
    NOT EQUAL VERIFY

    // check that k5 != d4
    <4> PICK
    <6> PICK
    NOT EQUAL VERIFY


// stack: <sd1> <d1> <sd2> <d2> <sd3> <d3> <sd4> <d4> <sd5> <d5> <k5> <k4> <k3> <k2> <k1>
// altstack:

2DROP 2DROP DROP
TOALT DROP
TOALT DROP
TOALT DROP
TOALT DROP
TOALT DROP

// stack:
// altstack: <d5> <d4> <d3> <d2> <d1>

// Whatever else


```